packages/flutter/lib/src/painting/inline_span.dart

```diff
@@ -0,0 +1,257 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui' as ui show ParagraphBuilder;
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/gestures.dart';
+import 'package:flutter/services.dart';
+
+import 'basic_types.dart';
+import 'text_style.dart';
+import 'text_painter.dart';
+
+/// An immutable span of inline content which form a paragraph.
+///
+/// The subclass [TextSpan] specifies text and may contain child [InlineSpan]s.
+///
+/// The subclass [WidgetSpan] specifies embedded inline widgets. Specify a
+/// widget by wrapping the widget with a [WidgetSpan]. The widget will be
+/// laid out inline within the paragraph.
+///
+/// Only [TextSpan] may contain children, so in order to include multiple
+/// [InlineSpan]s, the root node must be a [TextSpan]. Leaving the
+/// [TextSpan.text] field null results in the [TextSpan] acting as an empty
+/// node with a list of children.
+///
+/// {@tool sample}
+///
+/// This example shows a tree of [InlineSpan]s that make a query asking for a
+/// name with a [TextField] embedded inline.
+///
+/// ```dart
+/// RichText(
+///   text: TextSpan(
+///     text: 'My name is ',
+///     style: TextStyle(color: Colors.black),
+///     children: <InlineSpan>[
+///       WidgetSpan(
+///         alignment: InlineWidgetAlignment.baseline,
+///         baseline: TextBaseline.alphabetic,
+///         widget: ConstrainedBox(
+///           constraints: BoxConstraints(maxWidth: 100),
+///           child: TextField,
+///         )
+///       ),
+///       TextSpan(
+///         text: '.',
+///       ),
+///     ],
+///   ),
+/// )
+/// ```
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [Text], a widget for showing uniformly-styled text.
+///  * [RichText], a widget for finer control of text rendering.
+///  * [TextPainter], a class for painting [InlineSpan] objects on a [Canvas].
+@immutable
+abstract class InlineSpan extends DiagnosticableTree {
+  /// Creates a [InlineSpan] with the given values.
+  const InlineSpan({
+    this.style,
+    this.recognizer,
+    this.semanticsLabel,
+  });
+
+  /// The style to apply to this span.
+  ///
+  /// The [style] is also applied to any child spans when this is an instance
+  /// of [TextSpan].
+  final TextStyle style;
```

Comment:
is this relevant to placeholder spans?

Reply:
Certain properties such as decorations can still be applied to placeholders. This allows control over any decorations like underlines that may go from text->placeholder->text.